dali/python/nvidia/dali/ops.py

```diff
@@ -601,29 +601,108 @@ def _choose_device(inputs):
+def _is_integer_like(input):
+    if type(input) == int:
+        return True
+    if isinstance(input, Constant):
+        if input.dtype in _int_types:
+            return True
+    return False
+
+def _is_real_like(input):
+    if type(input) == float:
+        return True
+    if isinstance(input, Constant):
+        if input.dtype in _float_types:
+            return True
+    return False
+
+# <type> description required by ArithmeticGenericOp
+def _to_type_desc(input):
+    if  type(input) == int:
+        return "int32"
+    if type(input) == float:
+        return "float32" # TODO(klecki): current DALI limitation
+    if isinstance(input, Constant):
+        dtype_to_desc = {
+            DALIDataType.INT8: "int8",
+            DALIDataType.INT16: "int16",
+            DALIDataType.INT32: "int32",
+            DALIDataType.INT64: "int64",
+            DALIDataType.UINT8: "uint8",
+            DALIDataType.UINT16: "uint16",
+            DALIDataType.UINT32: "uint32",
+            DALIDataType.UINT64: "uint64",
+            DALIDataType.FLOAT16: "float16",
+            DALIDataType.FLOAT: "float32",
+            DALIDataType.FLOAT64: "float64",
+        }
+        return dtype_to_desc[input.dtype]
+    raise TypeError(("Constant argument to arithmetic operation not supported. Got {}, expected "
+            "a constant value of type 'int', 'float' or 'nvidia.dali.types.Constant'.")
+            .format(str(type(input))))
+
+
+# Group inputs into categories_idxs, _EdgeReferences, integer constants and real constants
+# The categories_idxs is a list that for an input `i` contains a tuple:
+# (category of ith input, index of ith input in appropriate category)
+def _group_inputs(inputs):
+    categories_idxs = []
+    edges = []
+    integers = []
+    reals = []
+    for input in inputs:
+        if isinstance(input, _EdgeReference):
+            categories_idxs.append(("edge", len(edges)))
+            edges.append(input)
+        elif _is_integer_like(input):
+            categories_idxs.append(("integer", len(integers)))
+            integers.append(input)
+        elif _is_real_like(input):
+            categories_idxs.append(("real", len(reals)))
+            reals.append(input)
         else:
-            dev_inputs = inputs
-        # Call it imediatelly
-        return op(*dev_inputs)
-
-
+            raise TypeError(("Argument to arithmetic operation not supported. Got {}, expected "
+                    "a return value from other DALI Operator  or a constant value of type 'int', "
+                    "'float' or 'nvidia.dali.types.Constant'.").format(str(type(input))))
+    if len(integers) == 0:
+        integers = None
+    if len(reals) == 0:
+        reals = None
+    return (categories_idxs, edges, integers, reals)
+
+# Generate the list of <input> subexpression as specified
+# by grammar for ArithmeticGenericOp
+def _generate_input_desc(categories_idx, integers, reals):
+    input_desc = ""
```

Comment:
nitpick idea: You could do `input_desc = []` and append strings to it. At the end you do `" ".join(input_desc)`

Reply:
Cool idea, will do it in a follow-up.